### PR TITLE
pm: device: runtime: API cleanup

### PR DESCRIPTION
**Pull request by gmarull**
While the device runtime API has been available for a long time, it has rarely been used (just a couple of drivers, so probably < 1%). I guess one of the reasons is the lack of documentation or usage guidelines. The aim of this RFC is to clarify API usage (e.g. why should I care about async?), make things a bit simpler and improve the documentation. 

Commit messages provide more details and context on the changes.

Other points to discuss:

- Should asynchronous be optional to save resources (ie force it to be sync depending on a Kconfig setting?)

### CODE REVIEW

**Review by yperess — COMMENTED** [review]
Looking at the device_runtime.c implementation I'm wondering if calling get/put should also require passing in a token that would make them effectively re-entrant. Basically, we wouldn't want one bad driver to call put more than get and put the device to sleep. If the caller had to pass a token we could make it re-entrant.

**Review by ceolin — APPROVED** [review]

**Review by teburd — COMMENTED** [review]
Looks good, waiting on the CI to approve

File: subsys/pm/device_runtime.c
Comment (gmarull) [reply subsys/pm/device_runtime.c]: adjusted

File: include/pm/device_runtime.h
Comment (gmarull) [reply include/pm/device_runtime.h]: latest commit moves the API to the `pm_device_runtime*` _namespace_

**Comment by JordanYates** [comment]
No complaints here, I use runtime power management extensively out-of-tree and have never found a use for `pm_device_get_async`.

**Review by JordanYates — COMMENTED** [review]
No complaints here, I use runtime power management extensively out-of-tree and have never found a use for pm_device_get_async.

File: include/pm/device_runtime.h
Comment (erwango) [inline include/pm/device_runtime.h:52]: I would reorder the functions effects here:
- First: "If the device is suspended, it will be resumed." > This is the primary effect expected by the caller
- Then: "This function will increase device usage count." > This is mostly interesting to the pm system

File: include/pm/device_runtime.h
Comment (erwango) [inline include/pm/device_runtime.h:70]: Ditto: "Suspend the device, based on usage count. Device usage count is decreased. Suspension happens if counter reaches zero"

File: include/pm/device_runtime.h
Comment (gmarull) [reply include/pm/device_runtime.h]: Adjusted documentation with a better description, let me know if it's better now

**Comment by gmarull** [comment]
note: also fixed issues with system PM tracing as CI was complaining

File: include/pm/device_runtime.h
Comment (erwango) [inline include/pm/device_runtime.h:98]: ditto

File: include/pm/device_runtime.h
Comment (erwango) [inline include/pm/device_runtime.h:76]: "down to 0" ;-)
Otherwise, this is fine to me

Comment (gmarull) [reply include/pm/device_runtime.h]: done

**Review by erwango — APPROVED** [review]
Approved (assuming the CI will eventually pass)

**Review by nashif — APPROVED** [review]

**Pull request merged**
